posts/spring2021/js-chats-9-typescript/index.md

```diff
@@ -0,0 +1,415 @@
+
+In this scenario, finding Alfred is trivially simple; although it doesn't throw an error, since Alfred sits in the global execution context, it's immediately invoked and the result is printed out to our console. Finding type-safety bugs in JavaScript is easy, right?
+
+Now, take a look at these snippets consuming the same hypothetical utility function, firstly in the context of a browser, and then in the context of an express endpoint.
+
+```jsx
+const button = document.querySeletor("#btn")
+const header = document.querySelector("#header")
+button.addEventListener("click", function() {
+  header.innerText = getUserGreeting(currentUser.firstName)
+		 + "\nWe've been trying to reach you about your vehicle's extended warranty"
+})
+```
+
+```jsx
+const router = new express.Router()
+router.get("/user-greeting", async function(req, res) {
+	const user = await userService.findUserById(req.query.user_id)
+	const greeting = getUserGreeting(user.firstName)
+  res.send(greeting)
+})
+module.exports = router
+```
+
+Unlike the example prior, the use of this utility function is event-driven. Why is this important? Because there aren't any immediate red flags that we've introduced a bug; finding this error now involves clicking a specific button, or testing a specific endpoint. Without integration tests for these click listeners and express endpoints, this bug suddenly isn't as noticeable as it used to be, and can very easily slip into production, hindering both user experience as well as developer productivity.
+
+Now, imagine this in the context of a large-scale project with hundreds of event listeners and with dozens of other branches consuming the same utility functions, services, middlewares etc.
+
+Suddenly, finding simple type-safety errors in development becomes non-trivial. In fact, without established integration testing, the flexibility that dynamic typing in JavaScript provides is often the culprit of why this class of bugs is so often pushed into production in large-scale JavaScript projects, with [up to 20%](https://earlbarr.com/publications/typestudy.pdf) of bugs being attributable to type safety alone.
+
+### TypeScript to the rescue?
+
+One tool that attempts to address these inherent type-safety issues is TypeScript, a language that both syntactically extends JavaScript and compiles to JavaScript.
+
+### What TypeScript is
+
+TypeScript is a static type-checker that analyzes your source code to catch silly type-safety errors in development before they become infuriating, costly runtime errors in production or during CI/CD tests.
+
+![https://i.stack.imgur.com/u3qk9.png](https://i.stack.imgur.com/u3qk9.png)
```

Comment:
let's download this image and put it in the repo as well. There are several benefits to this:
1. we can guarantee the image link will not expire 
2. Gatsby will do optimization (both lazy-loading and sizing) on local images 

In the description, put something descriptive for better accessibility. For instance, for this specific image, we can put `error output from tsc`